backend/crud.py

```diff
@@ -0,0 +1,211 @@
+from typing import Dict
+from sqlalchemy.orm import Session
+from sqlalchemy import select, update
+
+from .models import (
+    AttrType,
+    Attribute,
+    AttributeDefinition,
+    BoundFK,
+    Entity,
+    Schema,
+    Value,
+    val_map,
+    type_map
+)
+
+from .schemas import (
+    SchemaCreateSchema,
+    SchemaUpdateSchema,
+    AttributeCreateSchema
+)
+
+
+def create_attribute(db: Session, data: AttributeCreateSchema) -> Attribute:
+    exists = db.execute(select(Attribute).where(Attribute.name == data.name)).scalar()
+    if exists:
+        # TODO custom exceptions
+        raise Exception(f'Attribute with name `{data.name}` already exists')
+    a = Attribute(name=data.name, type=data.type)
+    db.add(a)
+    db.commit()
+    return a
+
+
+def create_schema(db: Session, data: SchemaCreateSchema) -> Schema:
+    # TODO require lock here and in update?
+    exists = db.execute(select(Schema).where( (Schema.name == data.name) | (Schema.slug == data.slug) )).scalar()
+    if exists:
+        raise Exception(f'Schema with name `{data.name}` or slug `{data.slug}` already exists')
```

Comment:
Yes, I agree. We don't want ambiguous names or slugs.

As I'm more familiar with the Django ORM I'm wondering if it would be sufficient to rely on the unique constraints, which are defined in the `Schema` model, causing an exception during the attempt to save the data?

Reply:
Well, we can go this way, but I'd rather prefer to raise custom exceptions and handle them correspondingly on API level

Reply:
Assuming that the SqlAlchemy ORM behaves like the Django ORM, maybe we can find a compromise:

We try saving the data to the database, catch the exception from SqlAlchemy and use the `raise ... from ...` syntax to get a custom exception on the API level for further handling?

Reply:
So, I checked that, we can wrap
```python
sch = Schema(name=data.name, slug=data.slug)
db.add(sch)
db.flush()
```
in try-except and catch `IntegrityError` from ORM and then raise our custom exception